Retrieve multiple tagged data slices in one go

jgrewe opened the pull request:
With this PR, it is possible to get several tagged slices with one call. In case the referenced DataArray has a RangeDimension, this should greatly improve the performance for it reduces the fileio dramatically. 

A few functions have been deprecated (9daaad6 ).

I changed the behavior of the indexOf function in RangeDimensions. Previously, we returned the index of the tick that is equal or greater than the position. Now it is less or equal, which in my view makes a lot more sense, so I tend to consider the previous behaviour a bug. Docs of Tag and MTag were updated to state this explicitly. Tests were adapted

gicmo commented:
Cool!! Created a bunch of [warnings](https://cdash.0xdeadbabe.info/index.php?project=nix) on windows though.

jgrewe commented:
even if it should compile, please do not yet merge. need to fix all warnings and will then rebase the branch

achilleas-k commented:
Nice change. Yes, this looks like a great idea, but I haven't looked through all the changes yet.
I'd also have to do the same on the NIXPY side.

jgrewe commented:
I am done with it :) I think

achilleas-k commented:
Gonna have to read through it on Monday

jgrewe commented:
@achilleas-k  take your time, it is larger, than I thought it would be.

jgrewe commented on src/Dimensions.cpp:
```diff
@@ -356,24 +386,58 @@ double RangeDimension::tickAt(const ndsize_t index) const {
+
+    for (size_t i = 0; i < (std::min(start_positions.size(), end_positions.size())); i++) {
+        if (start_positions[i] > end_positions[i] ) {
+            continue;
```
Well, yes, the loop only proceeds as long as we have start and end indices and if start is larger than end, we skip. You are perfectly right, that it would break the match between in- and output. I see only one alternative, i.e. throwing an exception, which would lead to runtime problems I wanted to avoid.

achilleas-k replied:
Yeah, I agree, throwing an exception wouldn't be nice.
What would happen if it instead returned -1 for each invalid position? Since the return values are indices, a negative index would signal an invalid value.

Not sure I like it entirely but I'd like to hear your thoughts.

jgrewe replied:
unfortunately, ndsize_t is unsigned and anything else would be wrong, methinks

achilleas-k replied:
Well there goes that idea. Yeah, didn't consider that.

gicmo replied:
There is nd**s**size_t in theory, but I am not sure that is a good use case here.

jgrewe replied:
na, don't think so. Even though it is unlikely that the full range of ndsize would ever be used in our context, I would rather not use it here. If you think it cleaner, I can change these situations to throw an exception in the cases that 1) the size of the vectors do not match, and 2) if end < start position.
Or, I can leave it as it is :)
In many cases this function is more or internal use, i.e. called via the data access functions of tag and mtag.

achilleas-k replied:
You make a good point about it being internal only. I'm wondering now how the situation could occur (that end < start). The start and end vectors are built from positions and extents, right?

jgrewe replied:
yes, indeed, but since it is part of the public API, there is, at least in principle, the opportunity for errors

gicmo commented:
@jgrewe can you create a new branch and cherry pick 9f42893f13f3eb9cb29d9415be418f02eff1de8b and file a PR with just that, so we get that travis fix separately?

jgrewe commented:
@gicmo sure :)

jgrewe commented:
I added a few commits. the following was done:

1. changed the behaviour of the indexOf function. It will complain when different numbers of start and end positions are passed.
2. add the option to pass an empty vector to as position_indices. In that case all slices will be returned.
3. fixed a bug for 1D data
4. added some tests

achilleas-k commented:
I didn't go through the whole thing in detail and it's hard to see exactly what changed since the last time I read through it, but I think I spotted all the latest changes.

I like it. LGTM
